### DISCUSSION TITLE
Kafka pods not running in different nodes with pod-antiaffinity specified

### DISCUSSION THREAD

**Discussion by shreyasarani23** [strimzi/strimzi-kafka-operator#4922] [General]
Below is my kafka configuration I have used pod antiaffinity to schedule kafka pods on different nodes
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-kafka-cluster
  labels:
    app: kafkaPod
spec:
  kafka:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
                - key: app
                  operator: In
                  values:
                  - kafkPod
            topologyKey: "kubernetes.io/hostname"
    version: 2.7.0
    replicas: 3
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      log.message.format.version: "2.7"
      auto.create.topics.enable: true
      auto.leader.rebalance.enable: true
      delete.topic.enable: true
    storage:
      type: persistent-claim
      size: 1Gi
      class: kafka-storage-class
      deleteClaim: true
  zookeeper:
    replicas: 3
    storage:
      type: persistent-claim
      size: 1Gi
      class: kafka-storage-class
      deleteClaim: true
  entityOperator:
    topicOperator: {}

But the problem is that only one kafka pod is getting deployed and the other two are in pending state.
When I describe the pod to check why it is in pending state it says the following 
0/3 nodes are available: didn't match pod affinity/anti-affinity.

Is there anything wrong in my deployment file?
Please correct me If I am wrong

**Comment by scholzj** [comment]
Please check the documentation for where the affinity should be configured. I don't think you have it in the right place.

**Reply by shreyasarani23** [reply]
I guess it is in right place 
This is from documentation:
apiVersion: kafka.strimzi.io/v1alpha1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    ...
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
                - key: application
                  operator: In
                  values:
                    - postgresql
                    - mongodb
          topologyKey: "kubernetes.io/hostname"
    ...
  zookeeper:
    ...

**Reply by scholzj** [reply]
Please share a link.

**Reply by shreyasarani23** [reply]
https://strimzi.io/blog/2018/07/30/running-kafka-on-dedicated-nodes/

**Reply by scholzj** [reply]
That is a 3 years old blog post. You need to check the documentation for the version you use: https://strimzi.io/documentation/

**Reply by shreyasarani23** [reply]
According to latest version this is the configuration they have used.
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
spec:
  kafka:
    # ...
    template:
      pod:
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              - labelSelector:
                  matchExpressions:
                    - key: application
                      operator: In
                      values:
                        - postgresql
                        - mongodb
                topologyKey: "kubernetes.io/hostname"
    # ...
  zookeeper:
    # ...

I will try out once with this

**Reply by shreyasarani23** [reply]
Hi I have tried to implement pod antiaffinity as per the documentation.
I have attached my kafka deployment file.
If there are any configuration mistakes please let me know.
[kafka-deployment.txt](https://github.com/strimzi/strimzi-kafka-operator/files/6442040/kafka-deployment.txt)
I am getting the following error
Events:
  Type     Reason            Age               From               Message
  ----     ------            ----              ----               -------
  Warning  FailedScheduling  8s (x3 over 94s)  default-scheduler  0/3 nodes are available: 3 node(s) didn't match pod affinity/anti-affinity, 3 node(s) didn't match pod anti-affinity rules.

**Reply by scholzj** [reply]
Well, the error means what it means ... you said that the Kafka pods should not run on the same node as other pods with the label `app: kafkaPod`. But you do not have enough nodes to to achieve this. So you either need to change the `requiredDuringScheduling` rule to `preferredDuringScheduling` or us a different label. Keep in mind, that they way you configured it ZooKeeper will have this label as well.

**Reply by shreyasarani23** [reply]
Ohk I got it So what should I do now If want to run kafka pods on different nodes?
How can I give different labels so that kafka pods don't run on same nodes?

**Reply by scholzj** [reply]
You need to use different label just for the Kafka brokers. You can use the `.spec.kafka.template...` section in the Kafka CR for  it. Or you can use one of the built-in Strimzi labels (do not know the name out of my head - you would need to check on the pods).

**Reply by shreyasarani23** [reply]
Can U share me the link where they have used different label for kafka brokers under .spec.kafka.template....
And how do u get the information about the in-built strimzi labels??

**Reply by scholzj** [reply]
I do not have the link at hand. But in the Using guide, search for _Customizing Kubernetes Resources_. The built-in pod labels are probably easiest to find by deploying the KAfka cluster without the affinity first.
